src/eval.ml

```diff
@@ -165,6 +172,36 @@ and eq_modulo : term -> term -> bool = fun a b ->
   let res = try eq_modulo [(a,b)]; true with Exit -> false in
   if !log_enabled then log_eqmd (r_or_g res "%a == %a") pp a pp b; res
 
+(** [tree_walk t s] tries to match stack [s] against tree [t] *)
+and tree_walk : Dtree.t -> stack -> (term * stack) option = fun itree istk ->
+  let depth = Dtree.iter (fun _ -> 1) (fun _ chd ->
+      let _, d = List.split chd in 1 + (List.extremum (>) d)) 0 itree in
+  let _ = Array.make depth TE_None in
+  (* Use above env in the tree walk *)
+  let rec walk : Dtree.t -> stack -> (Dtree.action * stack) option =
+    fun tree stk ->
+    match tree with
+      | Leaf(a)                                 -> Some(a, stk)
+      | Node({ swap = io ; children = ch ; _ }) ->
+        let nstk = match io with
+          | Some(i) -> List.swap_head stk i
+          | None    -> stk in
```

Comment:
Possible improvement: first copy the stack into an array and work on an array instead of a list.

Reply:
Then, it seems to me that swapping is not useful anymore, isn't it? The swap field being renamed into an index field (we however need to check that the index is valid).

Reply:
Array used since commit 8ab1af1